### PR TITLE
Autocomplete Improvements

### PR DESCRIPTION
**Pull request by FaultyFunctions**
Directly tackles #5 and includes various other improvements. Check out a7a1e55 for a fleshed out list of changes.

Let me know what you think, and if there are any issues. Thanks!

### CODE REVIEW

File: rt-shell/scripts/scr_shell_scripts/scr_shell_scripts.gml [hunk 1]
```diff
@@ -30,6 +30,10 @@ function sh_help() {
 	)
 }
 
+function sh_clear() {
+	obj_shell.clear();
+}
```
Comment (nickavv) [inline rt-shell/scripts/scr_shell_scripts/scr_shell_scripts.gml:35]: The `scr_shell_scripts` is just a set of examples, so adding `sh_clear` here won't get packaged and shipped with the bundle. I've been debating whether to have built-in functions, and if we should just declare them inside `obj_shell` as `sh_whatever` or if they should be done differently somehow. 

What are your thoughts?

**Review by nickavv — COMMENTED** [review]
At a glance I'm liking what I see, still need to go try it out later so I can provide feedback on the experience, and maybe look at the code in-editor as well. Will continue review later

File: rt-shell/scripts/scr_shell_scripts/scr_shell_scripts.gml (hunk 1 above)
Comment (FaultyFunctions) [reply rt-shell/scripts/scr_shell_scripts/scr_shell_scripts.gml:35]: Yeah I was thinking about that too. I think built in functions would be a good idea. I was thinking about how to accomplish user defined autocompletion for their own functions and we could do structs containing the data linked to each function or something like that. Then I was thinking we could have the user group these however they want so for example, they could have a development grouping and then they could have a player grouping. That way, the shell could be actually shipped with games and provide users with functionality, but keep them away from game breaking functionality meant only for development.

Native functions like "clear" and others could be a part of a "native" grouping. I have a naive approach to this working for my own game but it's hard coded and I'd like to flesh this out to be more dynamic. That would probably have to be it's own PR though.

Comment (nickavv) [reply rt-shell/scripts/scr_shell_scripts/scr_shell_scripts.gml:35]: Sounds like it's out of the scope of this PR. We'll leave it as-is for now, although maybe we'll fix it to have `clear` be a built-in function before we ship a release with this functionality.

**Comment by nickavv** [comment]
OK so I've played with it now and I've got some feedback on the user experience.

First off: the mouse stuff (scroll bars, clicking the suggestions) is awesome and works great, so good job on that

As far as the suggestion box, I really disagree with up/down arrows interacting with it. Especially if you have a couple commands in a row with no suggestions you can be using up arrow to go back through your command history, and as soon as you get to one with suggestions it starts cycling through those instead.

I'd suggest using tab/shift-tab to cycle through the suggestions instead. Tab already moves down the suggestion list, so might as well use shift-tab to go the other way, and then leave up/down free for history navigation.

Another thing, the shift-space to accept a recommendation thing is weird to me, maybe that's how it works in other consoles but it's really unintuitive in my mind. I'd rather see return/enter accept the recommendation (give the autosuggest box focus on that key while its open, rather than having it submit the command).

I'm sorry it's a lot, what are your thoughts on all that?

**Comment by FaultyFunctions** [comment]
What if we just make all these keybinds customizable? We'll leave the rt-shell way as default (with up and down scrolling through history) and then if people want to change the up and down arrows to scroll the suggestions they can change that. I think this would be the best of both worlds.

I agree that the shift+space thing is weird. That's def not how it works for other consoles, I just didn't know what to do there. I'll make the change to have enter respond to fill in the suggestion. Thanks!

**Comment by FaultyFunctions** [comment]
Okay I added native scripts to the shell object. These are defined in `user_event[0]` and look like this:

```gml
/// @desc closes the console
variable_global_set("sh_close", function() {
	close();
});
```

Basically it defines the function inside the global struct. Since these are defined before the shell adds the "sh_" functions to the autocomplete list, these are also included. Also since these are defined *within* the shell object, you don't need to call `obj_shell.close()`, you can just call `close()`.

What do you think of this solution, check it out and let me know. Thanks!

**Comment by nickavv** [comment]
I like the implementation for the native scripts, and the other changes you've made.

One thing, I don't think the theming scripts need to be native, I really had included them as examples of how a user could customize the shell to fit into their game, but as I encourage through the THEMING doc the idea is to set these properties once where you instantiate your shell rather than have them be live. I think it's unnecessary bloat to the final package's filesize.
In a similar vein I'm up in the air about the size/anchoring scripts being native, but I can see a value to it in that you may want to move the shell out of the way to see something you're working on with it. So I think I like the idea of keeping those native.

**Comment by nickavv** [comment]
Also a note on `clear`, when we insert lines to clear the screen. The clear command should still be added to the output, it should just be before the blank lines so it doesn't appear on-screen. But if you scroll up you ought to be able to see that clear was called!

**Comment by FaultyFunctions** [comment]
Okay this should be the final commit now unless there is something else you think needs to be changed or added. Give it a whirl and let me know what your thoughts!

**Comment by nickavv** [comment]
OK it's in a place that I'm happy with it now! Thanks for the great review back-and-forth, I think it's in a great place.

**Pull request merged**